### PR TITLE
Fix destruction of `ucxx::RequestTagMulti` and final status

### PR DESCRIPTION
**Pull request by pentschev**
This PR introduces various fixes:

* Fix destruction of `ucxx::RequestTagMulti`, which after moving to inheriting from `ucxx::Request`, using `setStatus()` was missing, thus causing the `ucxx::RequestTagMulti` to never cleanup its inflight request status. Calling `setStatus()` instead of writing the `_status` attribute directly resolves this.
* The status of `ucxx::RequestTagMulti` was until now always `UCS_INPROGRESS` or `UCS_OK` after completing. This is not right but there is no good way to combine the statuses of all underlying requests. Therefore we now set the status of the first failing request as the final status instead of `UCS_OK` if at least one of the requests failed. The user can still check each underlying's request status if granular information is required.
* Since `ucxx::RequestTagMulti` now inherits from `ucxx::Request`, the latter's trace logging for creation/destruction are sufficient, thus removing the redundant trace logs is appropriate.
* Add lifetime trace logs for `ucxx::BufferRequest`.
* Execute user-defined callback within the scope fo `ucxx::Request::setStatus()` to prevent accidental reordering mistakes, as the callback needs to always execute after the status is set, because the callback itself might make use of that information. Additionally it will prevent missing execution of the callback should `setStatus()` be called elsewhere in the future.
* Return `std::shared_ptr<Buffer> in allocateBuffer()` as returning raw pointers may cause leaks, and can be problematic specifically in `ucxx::RequestTagMulti` as the `Buffer*` will not be released unless the user takes ownership and releases it appropriately.
* Introduce mutex in `ucxx::Request` because of potential concurrency issues when running with the progress thread, which a mutex can resolve in `ucxx::Request`. A common case is when `ucxx::Endpoint` is registering an inflight request and `setStatus()` has just begun running, in which case the inflight request will be removed (a no-op as it hasn't been previously registered) and then actually registered, which will cause the `ucxx::Request` to never cleanup.

### CODE REVIEW

**Review by wence- — APPROVED** [review]
This looks right to me, thanks.

**Review by wence- — CHANGES_REQUESTED** [review]
Requesting changes, since I am not sure about two things:

1. Do we need a lock on the whole `Request` object for concurrent modification, or is it sufficient just to lock `_status`?
2. It looks like there are some `Request` calls that modify/read `_status` without a lock. Is this intentional?

**Comment by pentschev** [comment]
Thanks @wence- for another great review. You raised various valid concerns, I think I have addressed them all via https://github.com/rapidsai/ucxx/pull/72/commits/27d6fbdaf4ee07a0b2f871836b4936d78cb45c16, which now always protects `_status` and `_request` via a `std::recursive_mutex` (to ensure methods like `process()` can still call `setStatus()` without trying to lock twice a non-reentrant mutex twice in the same thread). Would you mind taking another look in case I missed something?

File: cpp/src/request.cpp [hunk 1]
```diff
@@ -167,35 +172,43 @@ void Request::process()
...
+  /**
+   * The callback may execute arbitrary user code (e.g., `getStatus()`) and that may require
+   * the mutex.
+   */
+  ucxx_trace_req_f(_ownerString.c_str(),
+                   _request,
+                   _operationName.c_str(),
+                   "callback %p",
+                   _callback.target<void (*)(void)>());
+  if (_callback) _callback(status, _callbackData);
```
[the remark this answers is not in the document]
Comment (pentschev) [reply cpp/src/request.cpp]: Now that the lock is reentrant we can actually hold the lock here, I changed it https://github.com/rapidsai/ucxx/pull/72/commits/c13e98a3cde483e26e1ab0bd97a0a3121458b95f.

File: cpp/src/request.cpp [hunk 2]
```diff
@@ -109,12 +120,11 @@ void Request::callback(void* request, ucs_status_t status)
...
   }
-  auto statusAttr = _status.load();
-  if (statusAttr != UCS_INPROGRESS)
+  if (_status != UCS_INPROGRESS)
     ucxx_trace("Request %p has status already set to %d (%s), callback setting %d (%s)",
                this,
-               statusAttr,
-               ucs_status_string(statusAttr),
+               _status,
+               ucs_status_string(_status),
```
[the remark this answers is not in the document]
Comment (pentschev) [reply cpp/src/request.cpp:127]: Good catch, that should now be resolved by https://github.com/rapidsai/ucxx/pull/72/commits/c13e98a3cde483e26e1ab0bd97a0a3121458b95f also.

**Review by wence- — APPROVED** [review]
Thanks Peter.

**Comment by pentschev** [comment]
@wence- I'm merging this in now, but it seems that the main issue with `test_send_recv_multi.py` was the [`rerun_on_failure` fixture](https://github.com/rapidsai/ucxx/blob/branch-0.33/python/ucxx/_lib_async/tests/conftest.py#L71) I added previously, it runs the underlying test but not the [`event_loop` fixture](https://github.com/rapidsai/ucxx/blob/0e3c8aa81765d856e18ba823b7162ba680c380a5/python/ucxx/_lib_async/tests/conftest.py#L40-L52), which is what was failing. Part of the docs mention that [pytest-rerunfailures closes the event loop before the `asyncio_timeout` fixture can run](https://github.com/rapidsai/ucxx/blob/0e3c8aa81765d856e18ba823b7162ba680c380a5/python/ucxx/_lib_async/tests/conftest.py#L64-L67), but right now I couldn't reproduce this. I'm not sure if I've done something differently in the past or if something changes in pytest-rerunfailures. In any case, I think https://github.com/rapidsai/ucxx/pull/72/commits/b0c20b9e5f2529b2aa1771392bbdee4c8baee45d improves things a bit, I'll keep an eye on whether this change will raise any future issues, if not I'll drop `rerun_on_failure` later.

**Pull request merged**